text/0068-entity-tags.md

```diff
@@ -0,0 +1,296 @@
+# Entity tags
+
+## Related issues and PRs
+
+- Reference Issues: [#305](https://github.com/cedar-policy/cedar/issues/305)
+- Supercedes [RFC 66](https://github.com/cedar-policy/rfcs/pull/66)
+- Implementation PR(s): (leave this empty)
+
+## Timeline
+
+- Start Date: 2024-05-29
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+This RFC proposes to extend the Cedar type system with the ability to attach _tags_ to entities. Entity tags are similar to records, where keys are like record attributes and values are like attribute values. But keys need not be enumerated in advance, as is required with record attributes, and all values must have the same type.
+
+The main body of this RFC proposes that tag keys must be _literals_ when used to look up a tag's value, just like record attributes, but Alternative A generalizes this proposal to allow keys to be dynamically computed. Alternative A thus adds expressive power but also implementation cost.
+
+### Basic example
+
+Here is a schema defining two entities, each of which contains entity tags.
+```
+entity User = {
+  jobLevel: Long,
+  authTags: Tags<Set<String>>,
+};
+entity Document = {
+  owner: User,
+  policyTags: Tags<Set<String>>, 
+};
+```
+The `User` and `Document` entities each have an attribute of type `Tags<Set<String>>`, identifying tags whose values are sets of strings. Here's a policy that conforms to this schema:
+```
+permit (
+  principal is User,
+  action == Action::"writeDoc",
+  resource is Document) 
+when {
+  document.owner == principal ||
+    (principal.jobLevel > 6 &&
+    resource.policyTags has "write" && 
+    principal.authTags has "write" && 
+    resource.policyTags["write"].containsAny(principal.authTags["write"]))
+};
+```
+This policy states that for a `User` to carry out the _writeDoc_ action on a `Document`, the user must own the document, or else the user's job level must be at least 6 and the document and the user must each have a `write` tag in their attached tags, where at least one of the user's write-tag's values must be present in the document's write-tag's values.
+
+## Detailed design
+
+Neither the Cedar evaluator/authorizer, the JSON format for entities, nor the JSON or natural syntax for policies needs to change to support entity tags. This is because tags are represented as records and support the same operations. Only the Cedar schema format, the validator, and the schema-based entity parser need to change to account for a new `Tags<T>` type.
+
+### Schema
+
+Tags have type `Tags<T>`, where `T` is the type of tag values. Only entity attributes can be given type `Tags<T>`, and `T` cannot directly mention another `Tags` type.
```

Comment:
Regarding indirectly referencing tags: based on a later example I believe indirect references through an entity type are allows, so `Tags<UserTags>` where `UserTags` has to be a entity and not a common type.

Reply:
Yes, normal records cannot include a `Tags` type as an attribute, per the schema grammar. Also per the grammar, you cannot write `Tags<Tags<Long>>` but you could write `Tags<Foo>` where `Foo` is an entity type that has a `Tags<Long>` attribute.

> Can common types be `Tags` types, 

This is not in the proposal, but we could add it. I left it out because of `Tags`' second-class nature -- you can only use them for entity attributes, so I worry it's misleading to define a general type for them.

> Is `Tags<T>` allowed, where T is (or contains) a common type;

Yes. Per the grammar the `T` is a normal `Type` nonterminal.

> we didn't clarify if normal sets can, i.e., if `Set<Tags<T>>`

The grammar precludes this: The `T` in `Set<T>` must be a `Type` non-terminal, and `Type` does not include `Tags<>` types.